pkg/image/trigger/annotations/annotations.go

```diff
@@ -0,0 +1,266 @@
+package annotations
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/runtime"
+
+	ometa "github.com/openshift/origin/pkg/api/meta"
+	triggerapi "github.com/openshift/origin/pkg/image/api/v1/trigger"
+	"github.com/openshift/origin/pkg/image/trigger"
+)
+
+const TriggerAnnotationKey = "alpha.image.openshift.io/triggers"
+
+func calculateAnnotationTriggers(m meta.Object, prefix string) (string, string, []triggerapi.ObjectFieldTrigger, error) {
+	var key, namespace string
+	if namespace = m.GetNamespace(); len(namespace) > 0 {
+		key = prefix + namespace + "/" + m.GetName()
+	} else {
+		key = prefix + m.GetName()
+	}
+	t, ok := m.GetAnnotations()[TriggerAnnotationKey]
+	if !ok {
+		return key, namespace, nil, nil
+	}
+	triggers := []triggerapi.ObjectFieldTrigger{}
+	if err := json.Unmarshal([]byte(t), &triggers); err != nil {
+		return key, namespace, nil, err
+	}
+	if hasDuplicateTriggers(triggers) {
+		return key, namespace, nil, fmt.Errorf("duplicate triggers are not allowed")
+	}
+	return key, namespace, triggers, nil
+}
+
+func hasDuplicateTriggers(triggers []triggerapi.ObjectFieldTrigger) bool {
```

Comment:
If we combine the validation in set triggers with some kind of annotation being added to an object when triggers validation fails it will benefit end users, definitely.